mcumgr: Fix build when CONFIG_THREAD_MONITOR=n

SebastianBoe opened the pull request:
See commit messages

SebastianBoe commented:
I am not familiar with this code.

It needs to be reviewed/confirmed that it makes sense to not invoke this function when THREAD_MONITOR is disabled.

I have not checked if this bug exists in more instances either.

Perhaps there are more places we need to ifdef on THREAD_MONITOR.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -119,13 +119,15 @@ os_mgmt_taskstat_read(struct mgmt_ctxt *ctxt)
 
 	/* Iterate the list of tasks, encoding each. */
 	for (task_idx = 0; ; task_idx++) {
+#ifdef CONFIG_THREAD_MONITOR
```
You are turning off part of code that decides to jump out of this loop (line 124-126). I think that taskstat makes no sense if `CONFIG_THREAD_MONITOR` is disabled to the issue should be solved within Kconfig where it would not be possible to select taskstat when the monitor is not compiler in.

SebastianBoe replied:
I've pushed a commit that uses depends on.

Unfortunately OS_MGMT_TASKSTAT is default y.

So I can't check every enabler of OS_MGMT_TASKSTAT and have them also enable THREAD_MONITOR. 

It's not clear to me if select or depends is the best strategy here.

de-nordic replied:
I think that this is a configuration issue when an option of which whole functionality depends on completely optional, and by default not included, features of a system is defaulted to y.
Nothing else depend on this config, except for user cases and an outside software that would query the data, so there is no reason for the option to be defaulted to y by default.
I think that we should remove the default y, add note that users now have to select the thing in prj and make compilation optional for entire unit that serves taskstat.

SebastianBoe replied:
I agree.

Features should be default n.

Done.

SebastianBoe replied:
This ifdef should be changed to ifdef OS_MGMT_TASKSTAT.

Right?


```
#ifdef CONFIG_THREAD_MONITOR
static const struct k_thread *
zephyr_os_mgmt_task_at(int idx)
{

```

SebastianBoe replied:
"make compilation optional for entire unit that serves taskstat."

I've changed the ifdef to be on TASKSTAT instead of THREAD_MONITOR.

But I'm not sure what you are asking for here.

We could move the taskstat functions into their own C file if you wish.

de-nordic replied:
Yes, you are right, The "most local" configuration option available should be used and in this case the `OS_MGMT_TASKSTAT` would be one, as the inclusion should depend on mcumgr configuration.

de-nordic replied:
Can you please also add commit adding release note here https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.0.rst#libraries--subsystems that now Taskstat will not be available for selection when thread monitor is disabled?

SebastianBoe replied:
Could you help me write the text? Then I can do the paperwork.

It's not clear to me how to phrase this, since AFAICT, it has never been possible to use TASKSTAT while THREAD_MONITOR was disabled.

SebastianBoe replied:
We should at least definitely have:

OS_MGMT_TASKSTAT is now disabled by default.

in the changelog

de-nordic replied:
> But I'm not sure what you are asking for here.
> 
> We could move the taskstat functions into their own C file if you wish.

You are probably referring to 

> (...)  as the inclusion should depend on mcumgr configuration.

my bad for not making this clear: I do not required you to move the code anywhere; the unfortunate misuse of "inclusion" is referring to pre-processor making the code available for further compilation as an opposite to excluding it from further processing.

de-nordic replied:
> Could you help me write the text? Then I can do the paperwork.
> 
> It's not clear to me how to phrase this, since AFAICT, it has never been possible to use TASKSTAT while THREAD_MONITOR was disabled.

OK, you are right. We do not have to make any addition to notes. Thanks for pointing that out.

SebastianBoe committed:
mcumgr: kconfig: Have OS_MGMT_TASKSTAT default to n

OS_MGMT_TASKSTAT depends on THREAD_MONITOR, so when it defaults to y
and THREAD_MONITOR is disabled we get a build error.

Add a depends on to enforce a correct configuration.

Remove default y, as features should be default n.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
mcumgr: kconfig: Have task funcs ifdef on CONFIG_OS_MGMT_TASKSTAT

zephyr_os_mgmt_task_at and friends are ifdef'ing on a feature they
depend on instead of their own Kconfig option.

Which is not correct.

ifdef on CONFIG_OS_MGMT_TASKSTAT instead.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

de-nordic commented:
One small request, If you do not mind: can you add commit that would modify the https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/mgmt/mcumgr/smp_svr/prj.conf smp_svr sample configuration to enable the taskstat by default?

SebastianBoe commented:
@de-nordic: Done

de-nordic reviewed (approved):
Looks OK. Thank you!
